### PR TITLE
EIP2384 - Ice Age Adustment around Istanbul

### PR DESCRIPTION
**Pull request by shemnon**
Add support for EIP2384

* Current placeholder is "EIP2384Block" because Icestanbul seemed a
  little too silly.
* Add difficulty calculator tests from the reference tests

(https://github.com/ethereum/EIPs/pull/2384)

Signed-off-by: Danno Ferrin \<danno.ferrin@gmail.com\>

### CODE REVIEW

**Comment by RatanRSur** [comment]
Icestanbul is an amazing name.

File: config/src/main/resources/ropsten.json [hunk 1]
```diff
@@ -9,6 +9,7 @@
     "constantinopleBlock": 4230000,
     "constantinopleFixBlock": 4939394,
     "istanbulBlock": 6485846,
+    "icetanbulBlock": 7117117,
```
Comment (RatanRSur) [inline config/src/main/resources/ropsten.json]: still icestanbul here, is that intentional?

Comment (shemnon) [reply config/src/main/resources/ropsten.json]: Nope, and a test should have caught that too.  Patches incoming.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java [hunk 2]
```diff
@@ -67,15 +69,20 @@ private MainnetDifficultyCalculators() {}
...
-  public static DifficultyCalculator<Void> CONSTANTINOPLE =
+  static DifficultyCalculator<Void> CONSTANTINOPLE =
       (time, parent, protocolContext) ->
-          calculateByzantiumDifficulty(time, parent, CONSTANTINOPLE_FAKE_BLOCK_OFFSET);
+          calculateThawedDifficulty(time, parent, CONSTANTINOPLE_FAKE_BLOCK_OFFSET);
 
-  private static BigInteger calculateByzantiumDifficulty(
+  static DifficultyCalculator<Void> MUIR_GLACIER =
+      (time, parent, protocolContext) ->
+          calculateThawedDifficulty(time, parent, MUIR_GLACIER_FAKE_BLOCK_OFFSET);
```
Comment (RatanRSur) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java]: Do these need `@VisibleForTesting`s too?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java [hunk 3]
```diff
@@ -67,15 +69,20 @@ private MainnetDifficultyCalculators() {}
...
+
+  private static BigInteger calculateThawedDifficulty(
```
Comment (RatanRSur) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java]: Could this be `calculateMuirGlacierDifficulty`? As it is it seems like it reduces code discoverability if you're searching for instances of the hard fork, like you could have when it was `Byzantium`?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java (hunk 2 above)
Comment (RatanRSur) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java]: Ah, I guess, not, found their uses.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java (hunk 3 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetDifficultyCalculators.java]: It's used in 3 different places, byzantium, constantinople, and now MuirGlacier.  Before it was `calculageByzantiumDifficulty`.

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/mainnet/DifficultyCalculatorTests.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        new Object[] {
+          "/BasicTests/difficultyEIP2384_random_to20M.json",
+          MainnetProtocolSchedule.fromConfig(
+              GenesisConfigFile.fromConfig("{\"config\":{\"muirGlacierBlock\":0}}")
+                  .getConfigOptions())
+        });
+  }
+
+  @Test
+  public void testDifficultyCalculation() throws IOException {
```
Comment (RatanRSur) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/mainnet/DifficultyCalculatorTests.java]: Can this test not be run in isolation? Getting NullPointerExceptions when running this alone through intellij. Is there a gradle task that downloads resources?

Comment (shemnon) [reply ethereum/core/src/test/java/org/hyperledger/besu/ethereum/mainnet/DifficultyCalculatorTests.java]: How many, if its 3 NPEs You need to update your submodules, if it's more you need to run it via gradle or have IntelliJ add the reference tests resoureces to the classpath when running.

File: config/src/test-support/java/org/hyperledger/besu/config/StubGenesisConfigOptions.java [hunk 5]
```diff
@@ -262,6 +269,11 @@ public StubGenesisConfigOptions istanbulBlock(final long blockNumber) {
     return this;
   }
 
+  public StubGenesisConfigOptions muirGlacierBlock(final long blockNumber) {
+    muirGlacierBlockNumber = OptionalLong.of(blockNumber);
+    return this;
+  }
```
Comment (RatanRSur) [inline config/src/test-support/java/org/hyperledger/besu/config/StubGenesisConfigOptions.java]: This is unused, does it need to be used in `ReferenceTestProtocolSchedules`?

Comment (shemnon) [reply config/src/test-support/java/org/hyperledger/besu/config/StubGenesisConfigOptions.java]: Yep.  Added.

**Pull request merged**